tensorflow/contrib/gdr/gdr_memory_manager.cc

```diff
@@ -557,96 +506,137 @@ void GdrMemoryManager::TensorFromTransportOptions(
           CreateEndpoint(remote_mr.host(), remote_mr.port(), iter->second);
       if (!s.ok()) {
         done(s);
+        if (host_copy) {
+          delete host_copy;
+        }
         return;
       }
     }
+    id = iter->second.get();
   }
-  rdma_cm_id* id = iter->second.get();
 
   uint64_t start = Env::Default()->NowMicros();
 
+  StatusCallback callback = [tensor, remote_mr, device, device_context, on_host,
+                             start, done, host_copy](const Status& s) {
+    if (!s.ok()) {
+      done(s);
+      if (host_copy) {
+        delete host_copy;
+      }
+      return;
+    }
+#if GOOGLE_CUDA
+    if (device->tensorflow_gpu_device_info() && !on_host && host_copy) {
+      GPUUtil::CopyCPUTensorToGPU(
+          host_copy, device_context, device, tensor,
+          [tensor, remote_mr, device, device_context, start, done,
+           host_copy](const Status& s) {
+            if (!s.ok()) {
+              done(s);
+              delete host_copy;
+              return;
+            }
+            uint64_t end = Env::Default()->NowMicros();
+
+            VLOG(2) << "RDMA from remote memory region " << remote_mr.rkey()
+                    << " of size " << DMAHelper::buffer(tensor)->size()
+                    << " with tensor key " << remote_mr.tensor_key() << " took "
+                    << (end - start) << " micros";
+
+            if (VLOG_IS_ON(2)) {
+              uint64_t checksum =
+                  GPUUtil::Checksum(device, device_context, *tensor);
+              CHECK(checksum == remote_mr.checksum())
+                  << "Checksum mismatch: " << checksum
+                  << "!=" << remote_mr.checksum();
+            }
+
+            done(Status::OK());
```

Comment:
FYI: Be aware that many GPUUtil functions, like CopyCPUTensorToGPU, execute their callbacks in the one polling thread of the EventMgr associated with the GPU.  It's the responsibility of the programmer to ensure that these callbacks are not blocking or otherwise long-running.  If they are, no further events will be processed for that GPU until the callback closure finally terminates which can delay other Ops.  If the callback needs to trigger expensive processing it must be moved to another thread.